### ISSUE TITLE
[Bug]: Excessive Redis Connections and waitUntilFinished Timeout in High-Load Environments

### ISSUE THREAD

**Issue by dbas-dn** [taskforcesh/bullmq#2749]
### Version

v5.12.12

### Platform

NodeJS

### What happened?

Environment:

Kubernetes: k3s with 3 nodes
Redis: Sentinel configuration with 3 nodes
Description:

Our system is designed to create a large number of queues. We noticed that while it's possible to pass an already established Redis connection for creating Queue, Job, and Worker instances, the QueueEvents class duplicates the passed connection using duplicate().

During our load tests, where we ran between 100-600 queues, each queue created jobs that returned results. We observed that under low load, waitUntilFinished worked as expected, returning the job result. However, under high load conditions, waitUntilFinished failed to return, hanging indefinitely until the TTL expired.

To debug this, I implemented a parallel polling mechanism that ran the scripts.isFinished script, which indicated that the job had indeed reached the completed status and I could retrieve its result, even though waitUntilFinished remained stuck.

Additionally, the excessive number of connections created by QueueEvents due to the duplicate() method led to us exceeding the maximum number of Redis connections, especially under high concurrency.

Workaround:

As a temporary workaround, I replaced waitUntilFinished with periodic execution of scripts.isFinished to check job completion.

Issue Summary:

waitUntilFinished does not return in high-load scenarios, even when the job is completed according to scripts.isFinished.
QueueEvents creates redundant connections by duplicating the Redis connection, which leads to exceeding the maximum number of connections under high concurrency.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by dbas-dn** [comment]
any reaction?

**Comment by manast** [comment]
```waitUntilFinished``` is not a recommended api to use in production, it does not scale and is not a proper way to design an architecture based on queues: https://blog.taskforce.sh/do-not-wait-for-your-jobs-to-complete/

**Comment by daimalou** [comment]
@manast I am facing the same. I am developing a realtime AI and Image convertion tool. The user use  `Server Sent Event` to push some image to my server, and i push to queue to process using third api(it has QPS) to convert and return to user until the job finished.

I think the `waitUntilFinished ` is very useful when using `Server Sent Event`.

**Comment by manast** [comment]
@daimalou it may be useful but it is not the proper way to use queues. As I see it, you would be better off just spawning a NodeJS worker thread, run the job and wait it for completion, than using a queue.

**Comment by manast** [comment]
@daimalou btw, SSE are used for sending data from the server to the client, not sure what you mean that you use it for sending images to the server :/. In any case, if you use SSE or web sockets, it does not matter, you can easily communicate to the client when a job has completed without relying on ```waitUntilFinished```, you may need to redesign your solution a bit, but thats the proper way to do it and have a scalable and issue system that run stable for a long time.

**Comment by daimalou** [comment]
@manast yes, my description was a bit unclear. i use https://github.com/Azure/fetch-event-source, it can post some date to server and server response data using SSE.
Thank you. I think what you said is correct. I need to create more logic myself.

**Comment by bencampion** [comment]
I recently had as similar sounding issue with `waitUntilFinished()` being flaky in a test. The test was doing something like this to wait for jobs to complete:

```js
const jobs = await queue.getJobs(['waiting', 'active']);
const queueEvents = new QueueEvents(queue.name, { connection });
await Promise.all(jobs.map((job) => job.waitUntilFinished(queueEvents)));
await queueEvents.close();
```

The test seemed always pass locally for me, but it was flaky in CI where the test runners have fewer resources . I could see from logging that the jobs were being processed successfully, but for some reason `waitUntilFinished()` wasn't always resolving and sometimes caused the test to timeout.

Adding `queueEvents.waitUntilReady()` to the test seams to have fixed the flakiness for me:

```js
const queueEvents = new QueueEvents(queue.name, { connection });
await queueEvents.waitUntilReady();
const jobs = await queue.getJobs(['waiting', 'active']);
await Promise.all(jobs.map((job) => job.waitUntilFinished(queueEvents)));
await queueEvents.close();
```

It looks like I might have had a race where I was using `QueueEvents` before it was ready to be used, but I've not investigated it any further to see if that is actually what was going on. I notice that the test case provided in this issue also doesn't call `waitUntilReady()`.
